src/OpenTK.Mathematics/Geometry/Ray3.cs

```diff
@@ -0,0 +1,143 @@
+﻿//
+// Ray.cs
+//
+// Copyright (C) 2021 OpenTK
+//
+// This software may be modified and distributed under the terms
+// of the MIT license. See the LICENSE file for details.
+//
+
+
+using System;
+using System.Diagnostics.Contracts;
+
+namespace OpenTK.Mathematics.Geometry {
+
+    /// <summary>
+    /// Defines a ray in 3d space.
+    /// </summary>
+    public struct Ray3 {
+
+        /// <summary>
+        /// Starting point of the ray.
+        /// </summary>
+        public Vector3 Origin { get; set; }
+
+        private Vector3 _direction { get; set; }
+
+        /// <summary>
+        /// Direction of the ray.
+        /// </summary>
+        public Vector3 Direction {
+            get => _direction;
+            set => _direction = value.Normalized();
+        }
+
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj) {
+            return obj is Ray3 ray && Equals(ray);
+        }
+
+        /// <inheritdoc/>
+        public bool Equals(Ray3 other) {
+            return Origin.Equals(other.Origin) &&
+                Direction.Equals(other.Direction);
+        }
+
+        /// <inheritdoc/>
+        public override int GetHashCode() {
+            return HashCode.Combine(Origin, Direction);
+        }
+
+        /// <inheritdoc/>
+        public override string ToString() {
+            return $"{Origin} -> {Direction}";
+        }
+
+        /// <param name="ray">The ray.</param>
+        /// <param name="distance">The distance to travel along the ray.</param>
+        /// <returns>The point <paramref name="distance"/> units along the ray. </returns>
+        [Pure]
+        public Vector3 GetPoint(float distance) {
+            return GetPoint(in this, distance);
+        }
+
+        /// <summary>
+        /// Get the closest point along the ray to a vector, with optional range clamping.
+        /// </summary>
+        /// <param name="ray">The ray</param>
+        /// <param name="vector">The vector</param>
+        /// <param name="nearClamp">Near clamp</param>
+        /// <param name="farClamp">Far clamp</param>
+        /// <returns>The closest point to the <paramref name="vector"/> along the <paramref name="ray"/>, within the clamped range.</returns>
+        [Pure]
+        public Vector3 GetClosestPoint(in Vector3 vector, float nearClamp = 0, float farClamp = float.PositiveInfinity) {
+            return GetClosestPoint(in this, in vector, nearClamp, farClamp);
+        }
+
+        /// <param name="ray">The ray.</param>
+        /// <param name="distance">The distance to travel along the ray.</param>
+        /// <returns>The point <paramref name="distance"/> units along the ray. </returns>
+        public static Vector3 GetPoint(in Ray3 ray, float distance) {
```

Comment:
This should have a version where you can `out` the point instead of having a return.